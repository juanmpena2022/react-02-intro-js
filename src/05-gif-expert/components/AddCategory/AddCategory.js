import PropTypes from "prop-types";
import { useState } from "react";

export const AddCategory = ({ setCategorias }) => {
  const [value, setValue] = useState("");

  //#region Funciones
  const iptChange = (evt) => {
    setValue(evt.target.value);
  };
  const submit = (evt) => {
    evt.preventDefault();
    
    let adicionado = false;    
    if (value.trim().length > 2) {
      setCategorias((cats) => {
        let result = [...cats];
        if (!cats.includes(value)) {
          adicionado = true;
          result.unshift(value);
        }

        return result;
      });

      if (adicionado) {
        setValue("");
      }
    }
  };
  //#endregion

  return (
    <form onSubmit={submit}>
      <input type="text" value={value} onChange={iptChange} />
    </form>
  );
};

AddCategory.propTypes = {
  setCategorias: PropTypes.func.isRequired,
};
