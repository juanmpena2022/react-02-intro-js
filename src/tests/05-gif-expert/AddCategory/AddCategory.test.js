import { shallow } from "enzyme";
import "@testing-library/jest-dom";
import { AddCategory } from "../../../05-gif-expert/components/AddCategory/AddCategory";

describe("pruebas del componente <AddCategory/>", () => {
  const setCategorias = jest.fn();
  let wrapper = shallow(<AddCategory setCategorias={setCategorias} />);

  beforeEach(() => {
    jest.clearAllMocks();
    wrapper = shallow(<AddCategory setCategorias={setCategorias} />);
  });

  test("snapshot", () => {
    expect(wrapper).toMatchSnapshot();
  });

  test("input de texto de cambiar", () => {
    const input = wrapper.find("input");
    const valor = "Hola Mundo";
    input.simulate("change", { target: { name: "width", value: valor } });

    const p = wrapper.find("p");
    const input1 = wrapper.find("input");

    expect(p.text()).toBe(valor);
    expect(input1.prop('value')).toBe(valor);
  });

  test("no debe postear info en el submit", () => {
    const f = wrapper.find("form");
    f.simulate("submit", { preventDefault() {} });

    expect(setCategorias).not.toHaveBeenCalled();
  });

  test("debe llamar setCategorias", () => {
    const input = wrapper.find("input");
    const valor = "Hola Mundo";
    input.simulate("change", { target: { name: "width", value: valor } });

    const f = wrapper.find("form");
    f.simulate("submit", { preventDefault() {} });

    expect(setCategorias).toHaveBeenCalled();
    expect(setCategorias).toHaveBeenCalledTimes(1);
    expect(setCategorias).toHaveBeenCalledWith(expect.any(Function));
  });
});
