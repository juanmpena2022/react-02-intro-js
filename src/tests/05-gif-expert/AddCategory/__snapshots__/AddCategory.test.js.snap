// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pruebas del componente <AddCategory/> snapshot 1`] = `
<form
  onSubmit={[Function]}
>
  <p />
  <input
    onChange={[Function]}
    type="text"
    value=""
  />
</form>
`;
